plugins/auth-backend/src/providers/pagerduty/provider.ts

```diff
@@ -0,0 +1,241 @@
+    {
+      fullProfile,
+      accessToken,
+      refreshToken,
+      params,
+    },
+    {
+      refreshToken,
+    },
+  );
+};
+export class PagerdutyAuthProvider implements OAuthHandlers {
+  private readonly _strategy: OAuth2Strategy;
+  private readonly logger: Logger;
+  private readonly catalogIdentityClient: CatalogIdentityClient;
+  private readonly tokenIssuer: TokenIssuer;
+  private readonly signInResolver?: SignInResolver<OAuthResult>;
+  private readonly authHandler: AuthHandler<OAuthResult>;
+
+  constructor(options: PagerdutyProviderOptions) {
+    this.logger = options.logger;
+    this.catalogIdentityClient = options.catalogIdentityClient;
+    this.tokenIssuer = options.tokenIssuer;
+    this.signInResolver = options.signInResolver;
+    this.authHandler = options.authHandler;
+    this._strategy = new OAuth2Strategy(
+      {
+        clientID: options.clientId,
+        clientSecret: options.clientSecret,
+        callbackURL: options.callbackUrl,
+        passReqToCallback: false,
+        authorizationURL: 'https://app.pagerduty.com/oauth/authorize',
+        tokenURL: 'https://app.pagerduty.com/oauth/token',
+      },
+      verifyFunction, // See the "Verify Callback" section
+    );
+  }
+
+  async start(req: OAuthStartRequest): Promise<RedirectInfo> {
+    return await executeRedirectStrategy(req, this._strategy, {
+      accessType: 'offline',
+      prompt: 'consent',
+      scope: req.scope,
+      state: encodeState(req.state),
+    });
+  }
+
+  async handler(req: express.Request) {
+    const { result, privateInfo } = await executeFrameHandlerStrategy<
+      OAuthResult,
+      any
+    >(req, this._strategy);
+
+    return {
+      response: await this.handleResult(result),
+      refreshToken: privateInfo.refreshToken,
+    };
+  }
+
+  private async handleResult(result: OAuthResult) {
+    const context = {
+      logger: this.logger,
+      catalogIdentityClient: this.catalogIdentityClient,
+      tokenIssuer: this.tokenIssuer,
+    };
+    const { profile } = await this.authHandler(result, context);
+
+    const response: OAuthResponse = {
+      providerInfo: {
+        idToken: result.params.id_token,
+        accessToken: result.accessToken,
+        scope: result.params.scope,
+        expiresInSeconds: result.params.expires_in,
+      },
+      profile,
+    };
+
+    if (this.signInResolver) {
+      response.backstageIdentity = await this.signInResolver(
+        {
+          result,
+          profile,
+        },
+        context,
+      );
+    }
+
+    return response;
+  }
+}
+
+export type PagerdutyAuthProviderOptions = {
+  authHandler?: AuthHandler<OAuthResult>;
+
+  signIn?: {
+    resolver?: SignInResolver<OAuthResult>;
+  };
+};
+
+export const pagerdutyDefaultSignInResolver: SignInResolver<
+  OAuthResult
+> = async (info, ctx) => {
+  const pagerduty = api({
+    token: info.result.accessToken,
+    tokenType: 'bearer',
+  });
+  const me = await pagerduty.get('/users/me');
```

Comment:
Curious what the difference is between `me` here and what you get back from the OAuth response? Generally we try to fetch at a minimum usable profile for the user within the handler itself, so if you don't get the user ID back in the auth result then let's move this fetch into the provider.